lib/master_manipulator/service.rb

```diff
@@ -2,71 +2,226 @@
+        # if its PE and not 3.8 use the status endpoint
+        # it its not PE or it is 3.8 use the simple curl call
+        # FIXME: Replace curl calls with http_request calls
+        if !pe_ver.match(three_eight_regex)
+          curl_call = "-I -k https://#{hostname}:8140"
+        else
+          curl_call = "-k -X GET -H 'Content-Type: application/json' https://#{hostname}:8140/status/v1/services?level=debug"
+        end
 
-        case @result.exit_code.to_s
+        (1..opts[:wait_cycles]).each do |i|
+          @result = curl_on(host, curl_call, :acceptable_exit_codes => [0,1,7])
+          # parse body if we are using PE and we are not in PE 3.8
+          (pe_ver && !pe_ver.match(three_eight_regex)) ? @body = JSON.parse(@result.stdout) : @body = []
+          
+          case @result.exit_code.to_s
           when '0'
             sleep 20
             pe_ver.match(/three_eight_regex/) ? return : (return if @body.all? { |k, v| v['state'] == 'running' })
           when '1', '7'
             # Exit code 7 is "connection refused"
             sleep (i**(1.2))
+          end
+        end
+
+        total_time = Time.now - start_time
+        
+        message = "Attempted to restart #{opts[:wait_cycles]} times, " +
+          "waited #{total_time} seconds total. \nHere is the status " +
+          "reported by the puppetserver:\n"
+        @body.each do |k,v|
+          message << "\n'#{k}' state: #{v['state']} "
         end
+        raise message
+      end
+    end
+
+    # Send SIGHUP to service_name on host and wait for reload to complete
+    # ==== Attributes
+    # *+host+ - the host that this should operate on
+    # *+service_name+ - the service to restart
+    #
+    # ==== Example
+    # hup_server(master, 'puppetserver')
+    # hup_server(host, 'pe-puppetserver')
+    def hup_server(host, service_name)
+      timeout = 30
+
+      # FIXME: This should be replaced either by invoking "puppetserver reload"
+      # after EZ-68 gets merged
+      pid = on(host, "pgrep -fo #{service_name}").stdout.strip
+      on(host, "kill -HUP #{pid}")
+      sleep timeout
+
+      # FIXME This should use the new status endpoint /status/v1/services?level=debug"
+      url = "https://#{host}:8140/puppet/v3/status"
+      cert = get_cert(host)
+      key = get_key(host)
+      response_code = "0"
+      sleeptime = 1
+      while response_code != "200" && timeout > 0
+        sleep sleeptime
+        begin
+          response = https_request(url, 'GET', cert, key)
+        rescue StandardError => e
+          expected_errors = [ EOFError, Errno::ECONNREFUSED, Errno::ECONNRESET,
+                              OpenSSL::SSL::SSLError ]
+          if !expected_errors.include?e.class
+            raise e
+          else
+            puts "Ignoring expected exception #{e}; server is restarting"
+          end
+        end
+        response_code = response.code unless response == nil
+        timeout = timeout - sleeptime
+      end      
```

Comment:
if i read this correctly, most of the above can be replaced with a call to beaker's `curl_with_retries`?